pkgcore/ebuild/atom.py

```diff
@@ -11,6 +11,7 @@
 __all__ = ("atom", "transitive_use_atom")
 
 import string
+from itertools import product
```

Comment:
Background context, I originally wanted to use itertools.product, pkgcore just didn't yet support those versions of python.

That said, unless I missed something, converting to itertools here just eliminates the OOM; however, this codepath should still effectively hang any consuming code due to the exponential explosion.

Take a look at https://github.com/pkgcore/pkgcheck/commit/3a5a8e2abe04bd543b305932f923072189eb159f for background context; that atom (which isn't uncommon) is 2^25 possible permutations.

Your usage of product *will* suppress the OOM, but it would allow any code that isn't protected to go down that expoentential pathway, generating each possible permutation.  For pmerge, the resolver likely will cache those results anyways, resulting in OOM.

That said, assuming this code passes tests, your version is a helluva lot fucking cleaner than mine, so merging on that alone is probably wise; I just want to be clear that it shifts the problem from one axis to another (or at least, that's what I'm reading in the code while doing a 5m drive by review).